### PR TITLE
feat(medusa,medusa-payment-stripe): Move database mutation from plugin to core

### PR DESCRIPTION
**Pull request by adrien2p**
**what**
The goal of that PR is to first refactor the payment provider and payment plugin to support the new API that removes the data mutation from within the plugin to be done by the core instead. In any case, this pr does not include the steps of the deeper refactoring. The last part will come in later pr.

**How**
- The payment plugin is now capable to handle both the deprecated and new API and the plugin works the same as it use to works.
- The mutation made by the plugin have been moved into the core as well as the subscriber
- The tests have been updated to reflect the changed
- Remove all new methods introduced by the payment collections
  - Mutualise types
  - Update provider and payment collection services
  - cleanup around all those refactoring including cleanup of the payment collection
  - refactor stripe payment plugin

FIXES CORE-887

### CODE REVIEW

**Comment by adrien2p** [comment]
@olivermrbl After testing it, it is still working and I only have 1 intent and one customer for a cart 👍 

Screenshot 2022-12-08 at 16 38 47
Screenshot 2022-12-08 at 16 38 58
Screenshot 2022-12-08 at 16 38 53

**Comment by adrien2p** [comment]
I still have some cleanup to do before review 😊

File: packages/medusa/src/services/payment-provider.ts
Comment (carlos-r-l-rodrigues) [inline packages/medusa/src/services/payment-provider.ts:755]: nice

File: packages/medusa-payment-stripe/src/helpers/stripe-base.js
Comment (carlos-r-l-rodrigues) [inline packages/medusa-payment-stripe/src/helpers/stripe-base.js:99]:
```suggestion
      return
```

**Review by carlos-r-l-rodrigues — APPROVED** [review]
LGTM

**Comment by adrien2p** [comment]
@srindom would you be able to review this one so I can continue the work :)

File: packages/medusa-payment-stripe/src/helpers/stripe-base.js
Comment (olivermrbl) [inline packages/medusa-payment-stripe/src/helpers/stripe-base.js:157]: comment: Also seems to be formatted strangely.

File: packages/medusa-payment-stripe/src/helpers/stripe-base.js
Comment (olivermrbl) [inline packages/medusa-payment-stripe/src/helpers/stripe-base.js:136]: thought: Think we should send `stripe_customer_id` here to avoid ambiguity

File: packages/medusa-payment-stripe/src/helpers/stripe-base.js
Comment (olivermrbl) [inline packages/medusa-payment-stripe/src/helpers/stripe-base.js:160]: comments: Same comments are for `createPayment`

File: packages/medusa-payment-stripe/src/helpers/stripe-base.js
Comment (olivermrbl) [inline packages/medusa-payment-stripe/src/helpers/stripe-base.js:176]: thought: Where do you get `collected_data` from? Don't think you've defined it anywhere

File: packages/medusa/src/interfaces/payment-service.ts
Comment (srindom) [inline packages/medusa/src/interfaces/payment-service.ts:36]: thought: I think `collected_data` is an odd name. The PaymentProvider responds with some data that it wants the core to store. I think `update_requests` would be more suitable to explain the interaction between the core and the plugin. So:

```ts
export type PaymentSessionResponse = {
  update_requests: { customer: Record<string, unknown> } // maybe even { customer_metadata: Record<string, unknown> } to make it very clear where the data will be stored
  session_data: Record<string, unknown>
}
```

Furthermore, I think we should revise the PaymentContext to be:

```ts
export type PaymentContext = {
  cart: {
    context: Record<string, unknown>
    id: string
    email: string
    shipping_address: Address | null
    billing_address: Address | null // added
    shipping_methods: ShippingMethod[]
  }
  currency_code: string
  amount: number
  resource_id?: string
  customer?: Customer // changed - we don't yet have a notion of collected data, and it is reasonable always to provide the customer if we know it
}
```

--

I think we have confused two separate ideas. Let me know if this makes sense to you?

File: packages/medusa/src/interfaces/payment-service.ts
Comment (srindom) [inline packages/medusa/src/interfaces/payment-service.ts:128]: thought: can we kill all the `*New` methods?

File: packages/medusa/src/interfaces/payment-service.ts
Comment (srindom) [inline packages/medusa/src/interfaces/payment-service.ts:51]: question: why is it `Cart & PaymentContext` should this be `PaymentContext` only?

File: packages/medusa/src/subscribers/cart.ts
Comment (olivermrbl) [inline packages/medusa/src/subscribers/cart.ts:38]: thought: Could this subscriber conflict with the call to create payment sessions i.e. the issue we currently have? 

Should it be serializable?

File: packages/medusa/src/services/payment-provider.ts
Comment (olivermrbl) [inline packages/medusa/src/services/payment-provider.ts:756]: praise: Nice control flow

File: packages/medusa/src/services/payment-provider.ts
Comment (srindom) [inline packages/medusa/src/services/payment-provider.ts:781]: thought: I would suggest that we never pass the entire cart to the payment provider plugin, to avoid provider plugins heavily relying on that data structure. What do you think of that?

File: packages/medusa/src/services/payment-provider.ts
Comment (srindom) [inline packages/medusa/src/services/payment-provider.ts:857]: question: this shouldn't override any existing metadata. Just to be sure - are we using `setMetadata` in the CustomerService?

File: packages/medusa/src/services/payment-provider.ts
Comment (olivermrbl) [inline packages/medusa/src/services/payment-provider.ts:854]: question: Should we call it something else than `customer` when it is collected from the payment provider to avoid ambiguity?

E.g. `provider_customer_id` or something similar

**Review by olivermrbl — CHANGES_REQUESTED** [review]

File: packages/medusa/src/interfaces/payment-service.ts
Comment (olivermrbl) [reply packages/medusa/src/interfaces/payment-service.ts:36]: ^ +1

Comment (adrien2p) [reply packages/medusa/src/interfaces/payment-service.ts:36]: I am alright to change the naming, i though if passing the collected data because the plugin consumer is not suppose to know where they are store. But know that I think about it, the doc will be able to say that they will always be in the metadata 😅 otherwise agree with you.

File: packages/medusa/src/interfaces/payment-service.ts
Comment (adrien2p) [reply packages/medusa/src/interfaces/payment-service.ts:51]: This is to prevent any breaking changes with the actual api. At the moment a cart is expected, this is why it is still a cart plus another type mixed. Once we decide to move forward and break the api, it will be removed

File: packages/medusa/src/services/payment-provider.ts
Comment (adrien2p) [reply packages/medusa/src/services/payment-provider.ts:857]: Yes we are and this is why i ended up just passing the object

File: packages/medusa/src/subscribers/cart.ts
Comment (adrien2p) [reply packages/medusa/src/subscribers/cart.ts:60]: I tried to concentrate on removing the mutation from the plugin in that pr and therefore not going to far. I have a separate pr which provide a new interface and abstract class with the new api which demonstrates what you are asking.

File: packages/medusa/src/services/payment-provider.ts
Comment (adrien2p) [reply packages/medusa/src/services/payment-provider.ts:756]: This is why I needed the cart for the actuals api to work

File: packages/medusa/src/services/payment-provider.ts
Comment (adrien2p) [reply packages/medusa/src/services/payment-provider.ts:223]: Already answered in a earlier comment 😅

File: packages/medusa-payment-stripe/src/helpers/stripe-base.js
Comment (adrien2p) [reply packages/medusa-payment-stripe/src/helpers/stripe-base.js:157]: It is not like that in my ide, i ll look into it. It must be some hidden chars

File: packages/medusa-payment-stripe/src/helpers/stripe-base.js
Comment (adrien2p) [reply packages/medusa-payment-stripe/src/helpers/stripe-base.js:136]: I could, but it would break any other existing payment plugins relying on stripe. Wdyt? A migration which duplicate the existing key with the new naming could help?

File: packages/medusa-payment-stripe/src/helpers/stripe-base.js
Comment (adrien2p) [reply packages/medusa-payment-stripe/src/helpers/stripe-base.js:176]: It is part of the type of object for the parameter

File: packages/medusa-payment-stripe/src/helpers/stripe-base.js
Comment (olivermrbl) [reply packages/medusa-payment-stripe/src/helpers/stripe-base.js:136]: Could we do `metadata?.stripe_id ?? metadata?.stripe_customer_id` ?

File: packages/medusa/src/subscribers/cart.ts
Comment (adrien2p) [reply packages/medusa/src/subscribers/cart.ts:38]: Yes sure, is it retiable?

File: packages/medusa-payment-stripe/src/helpers/stripe-base.js
Comment (olivermrbl) [reply packages/medusa-payment-stripe/src/helpers/stripe-base.js:176]: But you are not destructuring it on line 162 as I would expect, so it is `undefined` right?

File: packages/medusa-payment-stripe/src/helpers/stripe-base.js
Comment (adrien2p) [reply packages/medusa-payment-stripe/src/helpers/stripe-base.js:136]: We are not in control of the customisations or plugins made outside of medusa. It would be a breaking changes. I tried in that pr to have no breaking changes while introducing the new api. From the steps i have shared in the rfc, i proposed the breaking changes in a separate pr, when we will be ready to merge the interface and abstract class i ve proposed wdyt?

File: packages/medusa/src/services/payment-provider.ts
Comment (olivermrbl) [reply packages/medusa/src/services/payment-provider.ts:854]: This comment might be irrelevant if we choose to handle the collected data differently as per @srindom's comments

File: packages/medusa-payment-stripe/src/helpers/stripe-base.js
Comment (olivermrbl) [reply packages/medusa-payment-stripe/src/helpers/stripe-base.js:136]: Makes sense. Let's keep it as is :)

File: packages/medusa-payment-stripe/src/helpers/stripe-base.js
Comment (adrien2p) [reply packages/medusa-payment-stripe/src/helpers/stripe-base.js:176]: you are right, I ve mixed up the methods in my head, this one is the new and will be merged In the existing one instead. I ll fix it

File: packages/medusa/src/interfaces/payment-service.ts
Comment (adrien2p) [reply packages/medusa/src/interfaces/payment-service.ts]: included [here](https://github.com/medusajs/medusa/pull/2743/commits/5b30e2016c3ff51e2dc482d78f2f9f975a115326)

File: packages/medusa/src/interfaces/payment-service.ts
Comment (adrien2p) [reply packages/medusa/src/interfaces/payment-service.ts]: included [here](https://github.com/medusajs/medusa/pull/2743/commits/5b30e2016c3ff51e2dc482d78f2f9f975a115326)

File: packages/medusa/src/subscribers/cart.ts
Comment (adrien2p) [reply packages/medusa/src/subscribers/cart.ts:60]: included [here](https://github.com/medusajs/medusa/pull/2743/commits/5b30e2016c3ff51e2dc482d78f2f9f975a115326)

File: packages/medusa/src/services/payment-provider.ts
Comment (adrien2p) [reply packages/medusa/src/services/payment-provider.ts]: included [here](https://github.com/medusajs/medusa/pull/2743/commits/5b30e2016c3ff51e2dc482d78f2f9f975a115326)

File: packages/medusa-payment-stripe/src/helpers/stripe-base.js
Comment (adrien2p) [reply packages/medusa-payment-stripe/src/helpers/stripe-base.js]: included [here](https://github.com/medusajs/medusa/pull/2743/commits/5b30e2016c3ff51e2dc482d78f2f9f975a115326)

File: packages/medusa-payment-stripe/src/helpers/stripe-base.js
Comment (adrien2p) [reply packages/medusa-payment-stripe/src/helpers/stripe-base.js]: included [here](https://github.com/medusajs/medusa/pull/2743/commits/5b30e2016c3ff51e2dc482d78f2f9f975a115326)

File: packages/medusa-payment-stripe/src/helpers/stripe-base.js
Comment (adrien2p) [reply packages/medusa-payment-stripe/src/helpers/stripe-base.js]: included [here](https://github.com/medusajs/medusa/pull/2743/commits/5b30e2016c3ff51e2dc482d78f2f9f975a115326)

File: packages/medusa/src/interfaces/payment-service.ts
Comment (srindom) [reply packages/medusa/src/interfaces/payment-service.ts:50]: but doesn’t the signature overload ensure this? 

I don’t have experience with so just a question :)

Comment (adrien2p) [reply packages/medusa/src/interfaces/payment-service.ts:50]: No it doesn't, basically what it is doing is, allowing to pass either X data or Y data, then the method knowing that both are accepted needs to handle one or the other. But it does not take i to account that the implementation of the plugin have either been updated or not. Since the plugin can be updated and use the new data or can be the old version and still rely on the cart, we have to handle the changes in two steps in order to let the consumer than the old api will be deprecated. In that case in order to have no breaking changes both are merged. Then as explained in the comment of the new signature, the Cart is temporarily added and meant to disappear so that the consumer is aware that he should not use the cart data directly and only the new data 😅 does it make sense. 

As an example, i have refactored stripe but not the other plugins, but all plugins should still works. Even the ones not refactored and still relying on the cart

Comment (srindom) [reply packages/medusa/src/interfaces/payment-service.ts:50]: Got it makes sense!

**Review by patrick-medusajs — COMMENTED** [review]
I feel like I don't have enough domain knowledge to provide a meaningful review. That being said, I appreciate the approach taken in order to support a grace period before introducing a breaking change. I'm also seeing the good scout principle being applied "Leave this _codebase_ a little better than you found it."
Kudos 💪

File: packages/medusa-payment-stripe/src/helpers/stripe-base.js
Comment (adrien2p) [reply packages/medusa-payment-stripe/src/helpers/stripe-base.js:4]: It is part of the entity file at the moment, I think it should even be entirely moved to the types. wdyt?

File: packages/medusa-payment-stripe/src/helpers/stripe-base.js
Comment (adrien2p) [reply packages/medusa-payment-stripe/src/helpers/stripe-base.js]: Screenshot 2022-12-15 at 10 35 31

File: .eslintignore
Comment (adrien2p) [inline .eslintignore:11]: @srindom @olivermrbl in order to be able to lint and cleanup the tests I was working on, I had to remove these ignore. wdyt?

File: packages/medusa/src/migrations/1669032280563-payment-collection.ts
Comment (olivermrbl) [inline packages/medusa/src/migrations/1669032280563-payment-collection.ts:1]: question: Has this always been part of the PR? Seems misplaced

Comment (olivermrbl) [reply packages/medusa/src/migrations/1669032280563-payment-collection.ts:1]: Just saw your comment below. But should not be included now right?

File: packages/medusa/src/services/payment-provider.ts
Comment (olivermrbl) [inline packages/medusa/src/services/payment-provider.ts:218]: suggestion: Not entirely sure why this is necessary. Can I get you to add a comment for when this needs revisiting in the future?

File: packages/medusa/src/services/payment-provider.ts
Comment (olivermrbl) [inline packages/medusa/src/services/payment-provider.ts:703]: suggestion: This goes along with my other comment. Not entirely sure this condition + error message is comprehensible

**Review by olivermrbl — APPROVED** [review]
LGTM - pending approval from @carlos-r-l-rodrigues and @srindom 💪 

I think we need to revisit our payment provider documentation before this is released

File: packages/medusa/src/services/payment-provider.ts
Comment (adrien2p) [reply packages/medusa/src/services/payment-provider.ts:218]: Since we decided to merge the new method in that pr, it must take into account the payment collection. The amount on the payment session does not exists unless the order edit flag is enabled

File: packages/medusa/src/services/payment-provider.ts
Comment (adrien2p) [reply packages/medusa/src/services/payment-provider.ts:703]: Same as my previous comment

File: packages/medusa/src/migrations/1669032280563-payment-collection.ts
Comment (adrien2p) [reply packages/medusa/src/migrations/1669032280563-payment-collection.ts:1]: I need to re test, but i had an issue but in the end it was something else and I did not revert that one

File: packages/medusa/src/services/payment-provider.ts
Comment (olivermrbl) [inline packages/medusa/src/services/payment-provider.ts:701]:
```suggestion
       "Amount on payment sessions is only available with the OrderEditing API currently guarded by feature flag `MEDUSA_FF_ORDER_EDITING`.  Read more about feature flags here: https://docs.medusajs.com/advanced/backend/feature-flags/toggle/"
        "Unable to save the payment session with an amount. The feature flag order edit is not enabled."
```

**Comment by adrien2p** [comment]
> LGTM - pending approval from @carlos-r-l-rodrigues and @srindom 💪 
> 
> 
> 
> I think we need to revisit our payment provider documentation before this is released

The payment provider will probably evolve quickly in the next pr's. Once we are ready for the new api to take place, the methods signature will change very quickly.

File: packages/medusa/src/interfaces/payment-service.ts
Comment (srindom) [inline packages/medusa/src/interfaces/payment-service.ts:32]: suggestion (non-blocking): should we call this `customer_metadata` just to make it super clear what it is that the plugin can update?

File: packages/medusa/src/migrations/1669032280563-payment-collection.ts
Comment (srindom) [reply packages/medusa/src/migrations/1669032280563-payment-collection.ts:1]: todo: should be removed

**Review by srindom — CHANGES_REQUESTED** [review]
Almost ready :) 

A general comment:
I see that you moved `isDefined` to `medusa-core-utils` -- in the future I think we would benefit from splitting a cleanup like this into a separate PR. The clean-up changes create a lot of clutter, making the PR hard to review. No worries for now, but worth keeping in mind :)

File: packages/medusa/src/interfaces/payment-service.ts
Comment (adrien2p) [reply packages/medusa/src/interfaces/payment-service.ts:32]: Oli asked me to move it. I thought it was to include it here 😅

Comment (adrien2p) [reply packages/medusa/src/interfaces/payment-service.ts:32]: In fact, I might just have been a bit too fast. It was a thought from oli and i "thought" it was a request 😂

Comment (adrien2p) [reply packages/medusa/src/interfaces/payment-service.ts:32]: @srindom i ve miss placed the comments. It was about the isDefined 🤓

File: packages/medusa/src/interfaces/payment-service.ts
Comment (riqwan) [inline packages/medusa/src/interfaces/payment-service.ts:17]: suggestion: to provide more context to the naming, since we're defining a pattern to how plugins should be created. 

`${resource}${action}${Request | Response}`

Comment (riqwan) [inline packages/medusa/src/interfaces/payment-service.ts:17]: suggestion: to provide more context to the naming, since we're defining a pattern to how plugins should be created.

`${resource}${action}${Request | Response}`

File: packages/medusa/src/interfaces/payment-service.ts
Comment (riqwan) [inline packages/medusa/src/interfaces/payment-service.ts:140]: line 130 has the same method defined, wouldn't this method override that?

File: packages/medusa/src/models/payment-session.ts
Comment (riqwan) [inline packages/medusa/src/models/payment-session.ts:25]: question: what context will the cart be null for a payment session?

File: packages/medusa/src/subscribers/cart.ts
Comment (riqwan) [inline packages/medusa/src/subscribers/cart.ts:37]: future suggestion: we should split subscribers from handlers in the future, because:
1. Code modularisation
2. Allowing users to override or pass in custom handlers to subscribers.

File: packages/medusa/src/subscribers/cart.ts
Comment (riqwan) [inline packages/medusa/src/subscribers/cart.ts:62]: suggestion: you can remove await and async here since the services return a promise anyway.

**Review by riqwan — COMMENTED** [review]
Adding these as comments, feel free to discard them. Very cool PR!

File: packages/medusa/src/interfaces/payment-service.ts
Comment (adrien2p) [reply packages/medusa/src/interfaces/payment-service.ts:17]: This object is a context more than a request object. The consumer can use a part of fully the context or not at all eventually.

File: packages/medusa/src/interfaces/payment-service.ts
Comment (adrien2p) [reply packages/medusa/src/interfaces/payment-service.ts:140]: Since they have the same name, this is an overload, you can still implement one or the other but at least one is displayed as deprecated

Screenshot 2022-12-16 at 15 43 15

Comment (adrien2p) [reply packages/medusa/src/interfaces/payment-service.ts:140]: It can be seen roughtly like that

Screenshot 2022-12-16 at 15 44 35

File: packages/medusa/src/models/payment-session.ts
Comment (adrien2p) [reply packages/medusa/src/models/payment-session.ts:25]: In the case of the payment collection which is composed of resource id but not necessarily for the cart_id 😂 another concept ahah

File: packages/medusa/src/subscribers/cart.ts
Comment (adrien2p) [reply packages/medusa/src/subscribers/cart.ts:37]: I am not sure I understand what you are suggesting

File: packages/medusa/src/subscribers/cart.ts
Comment (adrien2p) [reply packages/medusa/src/subscribers/cart.ts]: You are right, but I will still keep the async as the linter enforce it

File: packages/medusa/src/services/payment-collection.ts
Comment (carlos-r-l-rodrigues) [inline packages/medusa/src/services/payment-collection.ts:397]: Is it possible to avoid sending this empty cart object here? We have the email in the customer object.

File: packages/medusa/src/interfaces/payment-service.ts
Comment (carlos-r-l-rodrigues) [inline packages/medusa/src/interfaces/payment-service.ts:18]: would make sense to make cart optional so we don't need it when creating a payment collection?

**Review by carlos-r-l-rodrigues — APPROVED** [review]
Just some small non-blocking comments, but overall it LGTM

File: packages/medusa/src/services/payment-collection.ts
Comment (adrien2p) [reply packages/medusa/src/services/payment-collection.ts:397]: I would like to keep that for the next pr which will be about the API definition around the interfaces and abstract class

File: packages/medusa/src/interfaces/payment-service.ts
Comment (adrien2p) [reply packages/medusa/src/interfaces/payment-service.ts:18]: same as above

**Review by srindom — APPROVED** [review]
LGTM!

**Pull request merged**
